### DISCUSSION TITLE
Manually executable TaskScheduler

### DISCUSSION THREAD

**Discussion by Denis535** [microsoft/vs-threading#1057] [Show and tell]
Default TaskScheduler just puts all the tasks into the thread pool.
But [for this reason](https://stackoverflow.com/questions/72707000/how-to-switch-forward-schedule-task-to-main-thread-if-i-use-game-loop-in-net) I need TaskScheduler which stores all my tasks until I will execute them manually.

Unfortunately, I did not find any solution. So, I tried to write it myself.
```
public class ManualTaskScheduler : TaskScheduler {

    private ConcurrentQueue<Task> Tasks { get; } = new ConcurrentQueue<Task>();
    public override int MaximumConcurrencyLevel => 1;

    public void Execute() {
        while (Tasks.TryDequeue( out var task )) {
            TryExecuteTask( task );
        }
    }
    protected override bool TryExecuteTaskInline(Task task, bool taskWasPreviouslyQueued) {
        return false;
    }

    protected override void QueueTask(Task task) {
        Tasks.Enqueue( task );
    }
    protected sealed override bool TryDequeue(Task task) {
        return false;
    }

    protected override IEnumerable<Task> GetScheduledTasks() {
        return Tasks.ToArray();
    }

}
```

Firstly, I want to ask you if I wrote it correctly? Or could there be a problem here?
Secondly, it would be nice if you added this to your library.

**Comment by Denis535** [comment]
Here is the version with locks and ManualResetEvent allowing you to wait for a task.
It looks like it works, but I'm afraid there might be bugs in some rare cases. Can you tell me if I did it right?

```
public class ManualTaskScheduler : TaskScheduler {

    private Queue<Task> Tasks { get; } = new Queue<Task>();
    private ManualResetEvent Event { get; } = new ManualResetEvent( false );
    public override int MaximumConcurrencyLevel => 1;


    public void Wait() {
        Event.WaitOne();
    }

    public void Execute() {
        lock (Tasks) {
            while (Tasks.TryDequeue( out var task )) {
                TryExecuteTask( task );
            }
            Event.Reset();
        }
    }
    protected override bool TryExecuteTaskInline(Task task, bool taskWasPreviouslyQueued) {
        return false;
    }

    protected override void QueueTask(Task task) {
        lock (Tasks) {
            Tasks.Enqueue( task );
            Event.Set();
        }
    }
    protected sealed override bool TryDequeue(Task task) {
        return false;
    }

    protected override IEnumerable<Task> GetScheduledTasks() {
        lock (Tasks) {
            return Tasks.ToArray();
        }
    }

}
```

**Reply by AArnott** [reply]
This repo probably isn't the best forum to be asking for general TaskScheduler advice. It really doesn't have to do with this library.

I'll say this though: your current implementation executes its tasks while holding a lock. That's generally a very bad idea and can lead to deadlocks, if for example one of the tasks wants to schedule another task with the same scheduler.

**Reply by Denis535** [reply]
Sorry.
I just thought it would be more reliable if my idea was implemented by professionals and available to many people.
And thanks for the advice.

**Reply by Denis535** [reply]
Now tasks are executed outside of lock.
```
    public void Execute() {
        var tasks = default( Task[] );
        lock (Tasks) {
            tasks = Tasks.ToArray();
            Tasks.Clear();
            Event.Reset();
        }
        var exceptions = default( List<Exception> );
        foreach (var task in tasks) {
            try {
                TryExecuteTask( task );
            } catch (Exception ex) {
                exceptions ??= new List<Exception>();
                exceptions.Add( ex );
            }
        }
        if (exceptions?.Count == 1) {
            throw exceptions[ 0 ];
        }
        if (exceptions?.Count >= 2) {
            throw new AggregateException( exceptions );
        }
    }
```

**Reply by AArnott** [reply]
I would recommend you throw `AggregateException` always. Otherwise your caller will have to deal with non-deterministic exception types being thrown based just on timing conditions around how many tasks are in the queue at the time they execute. 

It's also not clear to me what `Event` is for and what `Wait()` is expected to wait on. It seems it will wait until Execute is called (but unblock *before* those tasks execute), and block again as soon as the next task gets scheduled. If that's your intent, then the rest of your code loks accurate.

**Reply by Denis535** [reply]
Thank you!
It will wait until `QueueTask(task)` is called.
See example: https://stackoverflow.com/a/72718874/4805491

    public void Execute() {
        var tasks = default( Task[] );
        lock (Tasks) {
            tasks = Tasks.ToArray();
            Tasks.Clear();
            Event.Reset();
        }
        var exceptions = default( List<Exception> );
        foreach (var task in tasks) {
            try {
                TryExecuteTask( task );
            } catch (Exception ex) {
                exceptions ??= new List<Exception>();
                exceptions.Add( ex );
            }
        }
        if (exceptions != null && exceptions.Any()) {
            throw new AggregateException( exceptions );
        }
    }

**Reply by Denis535** [reply]
By the way, I did not implement the `TryExecuteTaskInline` and `TryDequeue` methods. Can it lead to problems or it's okay?

**Reply by AArnott** [reply]
It looks to me from your initial code snippet that you did indeed implement them (to return false). That should be fine. I mean, it'll have impact, but I think it's the impact you want.
